### PR TITLE
Separate RPC serialization from implementation

### PR DESCRIPTION
**Pull request by rphmeier**
This PR only does the `Eth` and `EthFilter` APIs but it was getting a bit big.

The `jsonrpc_core` update is for https://github.com/ethcore/jsonrpc-core/pull/27

### CODE REVIEW

File: rpc/src/v1/impls/eth.rs [hunk 1]
```diff
@@ -100,13 +99,13 @@ impl<C, S: ?Sized, M, EM> EthClient<C, S, M, EM> where
 		}
 	}
 
-	fn block(&self, id: BlockID, include_txs: bool) -> Result<Value, Error> {
+	fn block(&self, id: BlockID, include_txs: bool) -> Result<Option<Block>, Error> {
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: Shouldn't we have a generic `Error` also?

File: rpc/src/v1/impls/eth.rs [hunk 2]
```diff
@@ -271,256 +252,180 @@ impl<C, S: ?Sized, M, EM> Eth for EthClient<C, S, M, EM> where
...
-						Some(s) => Ok(to_value(&RpcH256::from(s))),
-						None => Err(errors::state_pruned()),
-					}
-				}
-			})
-
+	fn storage_at(&self, address: &Address, key: &H256, at: BlockNumber) -> Result<H256, Error> {
+		match at {
+			BlockNumber::Pending => Ok(take_weak!(self.miner).storage_at(&*take_weak!(self.client), address, key)),
+			id => match take_weak!(self.client).storage_at(address, key, id.into()) {
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: `.ok_or_else(|| errors::state_pruned())`

File: rpc/src/v1/impls/eth.rs [hunk 3]
```diff
@@ -271,256 +252,180 @@ impl<C, S: ?Sized, M, EM> Eth for EthClient<C, S, M, EM> where
...
-					id => match take_weak!(self.client).nonce(&address, id.into()) {
-						Some(nonce) => Ok(to_value(&RpcU256::from(nonce))),
-						None => Err(errors::state_pruned()),
-					}
-				}
-			})
+	fn transaction_count(&self, address: &Address, at: BlockNumber) -> Result<U256, Error> {
+		match at {
+			BlockNumber::Pending => Ok(take_weak!(self.miner).nonce(&*take_weak!(self.client), &address)),
+			id => match take_weak!(self.client).nonce(&address, id.into()) {
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: ^^^

File: rpc/src/v1/traits/eth.rs [hunk 4]
```diff
@@ -238,3 +385,386 @@ pub trait EthSigning: Sized + Send + Sync + 'static {
...
+		let version = try!(Eth::protocol_version(self));
+
+		Ok(Value::String(format!("{}", version)))
+	}
+
+	fn syncing(&self, params: Params) -> Result<Value, Error> {
+		try!(self.active());
+		try!(expect_no_params(params));
+
+		Eth::syncing(self).map(to_value)
```
Comment (tomusdrw) [inline rpc/src/v1/traits/eth.rs]: Wouldn't it be better to have different names instead?

`EthRpc::rpc_syncing` and call would be just `self.syncing()`?

File: rpc/src/v1/traits/eth.rs [hunk 5]
```diff
@@ -238,3 +385,386 @@ pub trait EthSigning: Sized + Send + Sync + 'static {
...
+	/// Sends signed transaction.
+	fn send_raw_transaction(&self, params: Params) -> Result<Value, Error> {
+		try!(self.active());
+
+		from_params::<(Bytes,)>(params).and_then(|(raw,)| {
+			let raw = raw.to_vec();
+			match UntrustedRlp::new(&raw).as_val() {
+				Ok(signed_transaction) => Eth::send_raw_transaction(self, signed_transaction)
+					.map(RpcH256::from).map(to_value),
+				Err(_) => Ok(to_value(RpcH256::from(H256::from(0)))),
```
Comment (tomusdrw) [inline rpc/src/v1/traits/eth.rs]: Could be a proper error instead (but it's out of scope of this PR)

File: rpc/src/v1/impls/eth.rs (hunk 1 above)
Comment (rphmeier) [inline rpc/src/v1/impls/eth.rs]: What would it consist of? The errors that can occur still need to be serialized and won't be used anywhere else.

File: rpc/src/v1/traits/eth.rs (hunk 4 above)
Comment (rphmeier) [inline rpc/src/v1/traits/eth.rs]: probably just personal taste, I wanted to make it clear that the `Eth` variant was being called.

File: rpc/src/v1/traits/eth.rs (hunk 5 above)
Comment (rphmeier) [inline rpc/src/v1/traits/eth.rs]: Agreed, this is just continuing the current behavior but an "invalid RLP" error would be nice.

**Comment by arkpar** [comment]
@debris please review this

**Comment by gavofyork** [comment]
this seems to increase the amount of effort to expose functionality by another function implementation per RPC. thus increase dev effort and maintenance burden.

i would prefer to reduce the effort, not increase it.

**Comment by tomusdrw** [comment]
But it keeps parameter deserialization in a single place in case of multiple implementations (which I suppose will be a case with light client) - keeping two APIs in sync in terms of expected parameters is easier.

What about a default implementation for `rpc_*` methods on the same trait?:

```rust
trait Eth {
 fn syncing(&self) -> Result<bool, Error>;
 fn rpc_syncing(&self, params: Params) -> Result<Value, Error> {
  try!(expect_no_params(params));
  self.syncing().map(to_value)
 }
}

**Comment by gavofyork** [comment]
tbh introducing a pattern that requires a new function `foo` to split some of its functionality into `x_foo` and call into that manually is pretty evil.

it's additional clutter which makes adding, renaming and removing functions more tedious than it really should be.

i'm just wondering whether it is something that can be fixed properly in due course (in which case i'd favour the q'n'd solution of just duplicating logic in order to minimise effort when adding functions and keep us reminded that it really does need addressing) or whether this is really the best we can reasonably do (in which case merging is reasonable).

**Comment by rphmeier** [comment]
Played around with macro and type-based codegen for parameter and output deserialization -- the results are here: https://github.com/ethcore/parity/commit/fd711c6c3613f1e8350e806df8c8259edf3bb97f

Only a single trait with strongly-typed methods is required for an RPC-API (with the caveat that all parameters must be deserializable except for a `BlockParam` which must be at the end) and are added to a delegate with `delegate.add_method("foo_Bar", move |base, params| foo_bar.wrap_rpc(base, params))`
which is the only added complexity. When `impl Trait` is stabilized even that can be elided.

@gavofyork think that this solution would be acceptable?

**Comment by gavofyork** [comment]
oh yeah! :)

**Comment by rphmeier** [comment]
Latest take on it is pushed -- takes a slightly different approach than described in the comment above because I was running into some issues with coercing anonymous function types into concrete function pointers without specifying the args explicitly.

RPC APIs are to be wrapped in the `build_rpc_trait!` macro, where you can just write out the definition like before.

Not supported yet: 
  - Async methods
  - Adding other attributes to RPC functions
  - named parameters

**Comment by gavofyork** [comment]
lgtm

**Pull request merged**